Feat kotlin android support

abnegate opened the pull request:
Adds Kotlin Android support. 

Generates an Android library module, which can be either:
- Directly imported into an Android project (useful for testing)
- Deployed to GitHub Packages with the included `deploy.gradle` file 
- Deployed to any maven repository by modifying the `repositories` block in `deploy.gradle`:
```groovy
    repositories {
        maven {
            name = "Your repo name"
            url = "Your repo url"
            credentials {
                username = System.getenv("YOUR_REPO_USERNAME")
                password = System.getenv("YOUR_REPO_PASSWORD")
            }
        }
    }
```

To run the deployment, from the project root, run:
```bash
    ./gradlew [module name]:assemble [module name]:publish
```


Where the module name will be the directory name of the generated SDK (default: kotlin-android)

Happy to make any changes required.

Closes #5

TorstenDittmann commented:
Thanks for your contribution! 🎉 

Awesome work, we will have this reviewed asap 👏🏻

christyjacob4 commented:
@abnegate You legend!!! This is amazing 🤩

abnegate commented:
@TorstenDittmann @christyjacob4 No problem, glad to help!

christyjacob4 commented on templates/kotlin/docs/example.md.twig:
```diff
@@ -0,0 +1,14 @@
+
+val client = {{ spec.title | caseUcfirst }}Client()
+{% if method.security|length > 0 %}
+  .setEndpoint('https://[HOSTNAME_OR_IP]/v1') # Your API Endpoint
```
```suggestion
  .setEndpoint("https://[HOSTNAME_OR_IP]/v1") // Your API Endpoint
```

christyjacob4 commented on templates/kotlin/docs/example.md.twig:
```diff
@@ -0,0 +1,14 @@
+  .setEndpoint('https://[HOSTNAME_OR_IP]/v1') # Your API Endpoint
+{% for node in method.security %}
+{% for key,header in node|keys %}
+  .set{{header | caseUcfirst}}('{{node[header]['x-appwrite']['demo']}}') # {{node[header].description}}
```
```suggestion
  .set{{header | caseUcfirst}}("{{node[header]['x-appwrite']['demo']}}") // {{node[header].description}}
```
Use double quotes for strings 
and // for comments 😄

christyjacob4 commented on templates/kotlin/src/main/java/io/appwrite/Client.kt.twig:
```diff
@@ -0,0 +1,220 @@
+            // Create an ssl socket factory with our all-trusting manager
+            val sslSocketFactory: SSLSocketFactory = sslContext.socketFactory
+            builder.sslSocketFactory(sslSocketFactory, trustAllCerts[0] as X509TrustManager)
+            builder.hostnameVerifier { _, _ -> true }
```
This was throwing an error in android studio. 

```suggestion
           builder.hostnameVerifier(HostnameVerifier { _ , _ -> true })
```

christyjacob4 commented:
<img width="1502" alt="Screenshot 2021-04-09 at 10 01 23 PM" src="https://user-images.githubusercontent.com/20852629/114212166-24122f80-997f-11eb-8d16-abf42e24a9c8.png">

@abnegate  Do you think this has something to do with the build.gradle in the library ?

christyjacob4 commented:
I did fix this by including the following in the app's gradle file 

```gradle
android {
  packagingOptions {
          exclude 'META-INF/LICENSE-notice.md'
          exclude 'META-INF/LICENSE.md'
      }
}
```

But is this due to the library ?

abnegate commented:
I can't reproduce by importing into a new project, are you able to post the app build.gradle? I've seen that error caused by dependency version collisions before

abnegate commented:
Could you try updating `androidx.core:core-ktx` to 1.3.2 and re-checking without the packaging options? The library exposes that package at that version which may cause the conflict

christyjacob4 commented:
<img width="797" alt="Screenshot 2021-04-09 at 10 39 57 PM" src="https://user-images.githubusercontent.com/20852629/114216422-8588cd00-9984-11eb-9c59-5314918e671f.png">

@abnegate  , I tried that . Doesn't seem to work

abnegate commented:
@christyjacob4 Got this sorted, it was caused by the test packages included in the library

christyjacob4 commented:
Okay Perfect 👌   @abnegate 
Still doing some testing in a demo app I'm building. 
Looks good so far. 
Cant stress the significance of this feature!

abnegate commented:
@christyjacob4 Sweet glad it's looking good! 

Extra bonus, it's all Java (Android at least) compatible. I'll run some tests with a plain java project to check without Android. In working on the Java templates, I realised that I'd need to generate overloads for all possible required+optional parameter permutations because of Java not supporting defaults so this is a big win.

With the Kotlin templates, if I simply add `@JvmOverloads` to the service method template, all the overloads will be generated based on the default params used in Kotlin:

<img width="1004" alt="Screen Shot 2021-04-10 at 6 03 32 PM" src="https://user-images.githubusercontent.com/5857008/114260146-229c4200-9a27-11eb-9aa6-53fac6543126.png">

christyjacob4 commented:
Okay that's a big win 🙌👍

christyjacob4 commented on templates/kotlin/src/main/java/io/appwrite/Client.kt.twig:
```diff
@@ -0,0 +1,221 @@
+                if (it.value is List<*>) {
+                    httpBuilder.addQueryParameter(it.key + "[]", it.value.toString())
+                } else {
+                    httpBuilder.addQueryParameter(it.key, it.value.toString())
```
@abnegate  Here's another thing I noticed in case of `DatabaseService.listDocuments()`
When we don't pass filters, the default argument is `null`

```kotlin
suspend fun listDocuments(
		collectionId: String,
		filters: List<Any>? = null,
		limit: Int = 25,
		offset: Int = 0,
		orderField: String = "",
		orderType: OrderType = OrderType.ASC,
		orderCast: String = "string",
		search: String = ""
	): Response
` ``
When we make the `GET` call using `client.call()`,  filters stays nulll and results in a URL like this 
```
https://demo.appwrite.io/v1/database/collections/60632e9bb9631/documents?filters=null&limit=25&offset=0&orderField=&orderType=ASC&orderCast=string&search=
```

`filters=null` causes a Server Side 500 error. 
 
Let's replace it so that only the `non-null params` in the hashmap are used in the query parameters. If there is a better way to do it, I'm welcome to that as well 😄  

```kotlin
 it.value?.let { it2 ->
                    if (it2 is List<*>) {
                        httpBuilder.addQueryParameter(it.key + "[]", it.value.toString())
                    } else {
                        httpBuilder.addQueryParameter(it.key, it.value.toString())
                    }
              }
```

christyjacob4 commented:
@abnegate  I'm also curious to know how you plan to simulate Coroutines in the Java

abnegate commented:
@christyjacob4 Null query params fixed.

Coroutines can be called from Java, an extra `Continuation` callback parameter is added to Kotlin suspending functions, my previous comment shows an example of how it'd work. 

Note that `resumeWith` may be called with an exception parameter, my example just hardcodes Response)

christyjacob4 commented:
Okay @abnegate thanks for that. Also wanted to check about the OAuth2 stuff.. We would have to open a web view like we do In flutter. The rest will be taken care of by the cookie jar

christyjacob4 commented:
@abnegate  Please also take a look at the `getFileDownload()` `getFilePreview()` and `getFileView()` functions. 
```kotlin
@Throws(AppwriteException::class)
    suspend fun getFileDownload(
		fileId: String
	): String {
        val path = "/storage/files/{fileId}/download".replace("{fileId}", fileId)
        val params = mapOf<String, Any?>(
            "project" to client.config["project"],
            "key" to client.config["key"]
        )

        val httpBuilder = HttpUrl.Builder().build().newBuilder(client.endPoint + path)
            ?: throw AppwriteException("Badly formed url!")

        params.forEach {
            httpBuilder.addQueryParameter(it.key, it.value.toString())
        }

        return httpBuilder.build().toString()
    }
```
Currently they return just a query string, when infact they are supposed to make a request to the server as in the case of flutter https://github.com/appwrite/sdk-for-flutter/blob/master/lib/services/storage.dart#L125

Please use this as a reference 
https://github.com/appwrite/sdk-generator/blob/master/templates/flutter/lib/services/service.dart.twig#L89-L94

abnegate commented:
@christyjacob4 Having a couple of issues getting these two sorted,

1. For OAuth, I think I'm missing something, everything seems to be working but there's no cookie coming back, instead the uri contains a fragment `#_=_`. Here is the flow I'm following if anything is off. It's the same as flutter as far as I can tell, only difference is the callback activity intent filter contains a host as well a scheme (seemed to be required, otherwise my appwrite instance would complain about a malformed redirect url because the host wasn't either appwrite or my own domain)

    - App opens webview to appwrite instance oauth with success + failure callback urls as query params
    - Appwrite redirects to provider log in
    - User logs in to provider portal
    - Provider redirects to appwrite instance with code
    - Appwrite instance redirects to app with key, secret parameters
    - App reads params in callback activity
    - App runs auth callback passing the full url containing cookie params
    - App builds cookie from params and stores in client cooke jar

2. I've fixed the return type of the "location" feature methods but I haven't been able to test them, I can't upload because I can't get clamav running either locally or on an AWS instance:

```
appwrite-clamav                 | Starting initial definition download
appwrite-clamav                 | Tue Apr 13 11:20:05 2021 -> ClamAV update process started at Tue Apr 13 11:20:05 2021
appwrite-clamav                 | Tue Apr 13 11:20:05 2021 -> ^Your ClamAV installation is OUTDATED!
appwrite-clamav                 | Tue Apr 13 11:20:05 2021 -> ^Local version: 0.102.4 Recommended version: 0.103.2
appwrite-clamav                 | Tue Apr 13 11:20:05 2021 -> DON'T PANIC! Read https://www.clamav.net/documents/upgrading-clamav
appwrite-clamav                 | Tue Apr 13 11:20:05 2021 -> daily.cvd database is up to date (version: 26138, sigs: 3970005, f-level: 63, builder: raynman)
appwrite-clamav                 | Tue Apr 13 11:20:05 2021 -> main.cvd database is up to date (version: 59, sigs: 4564902, f-level: 60, builder: sigmgr)
appwrite-clamav                 | Tue Apr 13 11:20:05 2021 -> bytecode.cvd database is up to date (version: 333, sigs: 92, f-level: 63, builder: awillia2)
appwrite-clamav                 | Tue Apr 13 11:20:05 2021 -> safebrowsing database available for download (remote version: 49191)
appwrite-clamav                 | Tue Apr 13 11:20:05 2021 -> ^downloadFile: Unexpected response (403) from https://database.clamav.net/safebrowsing.cvd
appwrite-clamav                 | Tue Apr 13 11:20:05 2021 -> ^getcvd: Can't download safebrowsing.cvd from https://database.clamav.net/safebrowsing.cvd
appwrite-clamav                 | Tue Apr 13 11:20:05 2021 -> Trying again in 5 secs...
```

I've added WIP commits if you want to test. For OAuth2 you'll need to add this to your `AndroidManifest.xml`:
```xml
        <activity android:name="io.appwrite.views.CallbackActivity" >
            <intent-filter android:label="android_web_auth">
                <action android:name="android.intent.action.VIEW" />
                <category android:name="android.intent.category.DEFAULT" />
                <category android:name="android.intent.category.BROWSABLE" />
                <data
                    android:scheme="callback-[PROJECT_ID]"
                    android:host="appwrite.io" />
            </intent-filter>
        </activity>
```
And this to your root `build.gradle` (I will find a way around this requirement but haven't found a suitable replacement yet):
```groovy
allprojects {
    repositories {
        maven { url "https://jitpack.io" }
    }
}
```
So your success/failure url's would be like `"callback-[PROJECT_ID]://appwrite.io"`

I'm still looking into both but appreciate any input you have!

lohanidamodar commented:
> @christyjacob4 Having a couple of issues getting these two sorted,
> 
> 1. For OAuth, I think I'm missing something, everything seems to be working but there's no cookie coming back, instead the uri contains a fragment `#_=_`. Here is the flow I'm following if anything is off. It's the same as flutter as far as I can tell, only difference is the callback activity intent filter contains a host as well a scheme (seemed to be required, otherwise my appwrite instance would complain about a malformed redirect url because the host wasn't either appwrite or my own domain)
>    
>    * App opens webview to appwrite instance oauth with success + failure callback urls as query params
>    * Appwrite redirects to provider log in
>    * User logs in to provider portal
>    * Provider redirects to appwrite instance with code
>    * Appwrite instance redirects to app with key, secret parameters
>    * App reads params in callback activity
>    * App runs auth callback passing the full url containing cookie params
>    * App builds cookie from params and stores in client cooke jar
> 2. I've fixed the return type of the "location" feature methods but I haven't been able to test them, I can't upload because I can't get clamav running either locally or on an AWS instance:
> 
> ```
> appwrite-clamav                 | Starting initial definition download
> appwrite-clamav                 | Tue Apr 13 11:20:05 2021 -> ClamAV update process started at Tue Apr 13 11:20:05 2021
> appwrite-clamav                 | Tue Apr 13 11:20:05 2021 -> ^Your ClamAV installation is OUTDATED!
> appwrite-clamav                 | Tue Apr 13 11:20:05 2021 -> ^Local version: 0.102.4 Recommended version: 0.103.2
> appwrite-clamav                 | Tue Apr 13 11:20:05 2021 -> DON'T PANIC! Read https://www.clamav.net/documents/upgrading-clamav
> appwrite-clamav                 | Tue Apr 13 11:20:05 2021 -> daily.cvd database is up to date (version: 26138, sigs: 3970005, f-level: 63, builder: raynman)
> appwrite-clamav                 | Tue Apr 13 11:20:05 2021 -> main.cvd database is up to date (version: 59, sigs: 4564902, f-level: 60, builder: sigmgr)
> appwrite-clamav                 | Tue Apr 13 11:20:05 2021 -> bytecode.cvd database is up to date (version: 333, sigs: 92, f-level: 63, builder: awillia2)
> appwrite-clamav                 | Tue Apr 13 11:20:05 2021 -> safebrowsing database available for download (remote version: 49191)
> appwrite-clamav                 | Tue Apr 13 11:20:05 2021 -> ^downloadFile: Unexpected response (403) from https://database.clamav.net/safebrowsing.cvd
> appwrite-clamav                 | Tue Apr 13 11:20:05 2021 -> ^getcvd: Can't download safebrowsing.cvd from https://database.clamav.net/safebrowsing.cvd
> appwrite-clamav                 | Tue Apr 13 11:20:05 2021 -> Trying again in 5 secs...
> ```
> 
> I've added WIP commits if you want to test. For OAuth2 you'll need to add this to your `AndroidManifest.xml`:
> 
> ```
>         <activity android:name="io.appwrite.views.CallbackActivity" >
>             <intent-filter android:label="android_web_auth">
>                 <action android:name="android.intent.action.VIEW" />
>                 <category android:name="android.intent.category.DEFAULT" />
>                 <category android:name="android.intent.category.BROWSABLE" />
>                 <data
>                     android:scheme="callback-[PROJECT_ID]"
>                     android:host="appwrite.io" />
>             </intent-filter>
>         </activity>
> ```
> 
> And this to your root `build.gradle` (I will find a way around this requirement but haven't found a suitable replacement yet):
> 
> ```groovy
> allprojects {
>     repositories {
>         maven { url "https://jitpack.io" }
>     }
> }
> ```
> 
> So your success/failure url's would be like `"callback-[PROJECT_ID]://appwrite.io"`
> 
> I'm still looking into both but appreciate any input you have!

@abnegate Great work. Loads of awesome stuff. 
Lately there's been some issues with ClamAV update server, so you can disable the ClamAV service in your installation. You can check the tutorial here for how to correctly disable ClamAV https://dev.to/appwrite/learn-how-to-disable-clamav-in-your-appwrite-stack-and-reduce-memory-usage-2e37

christyjacob4 commented on templates/kotlin/src/main/java/io/appwrite/Client.kt.twig:
```diff
@@ -0,0 +1,232 @@
+            return awaitResponse(request)
+        }
+
+        val body = if (MultipartBody.FORM.toString() == headers["content-type"]) {
+            val builder = MultipartBody.Builder().setType(MultipartBody.FORM)
+
+            params.forEach {
+                when {
+                    it.key == "file" -> {
+                        val file = it.value as File
+                        builder.addFormDataPart(it.key, file.name, file.asRequestBody())
+                    }
+                    it.value is List<*> -> {
+                        val list = it.value as List<*>
+                        for (index in list.indices) {
+                            builder.addFormDataPart(
+                                "${it.key}[${index}]",
+                                list[index].toString()
+                            )
+                        }
+                    }
+                    else -> {
+                        builder.addFormDataPart(it.key, it.value.toString())
+                    }
+                }
+            }
+            builder.build()
+        } else {
```
```suggestion
val body = if (MultipartBody.FORM.toString() == headers["content-type"]) {
            val builder = MultipartBody.Builder().setType(MultipartBody.FORM)
            params.forEach {
                when {
                    it.key == "file" -> {
                        val file = it.value as File
                        builder.addFormDataPart(it.key, file.name, file.asRequestBody())
                    }
                    it.value is List<*> -> {
                        val list = it.value as List<*>
                        for (index in list.indices) {
                            builder.addFormDataPart(
                                "${it.key}[]",
                                list[index].toString()
                            )
                        }
                    }
                    else -> {
                        builder.addFormDataPart(it.key, it.value.toString())
                    }
                }
            }
            builder.build()
        } else {
```

@abnegate  while reviewing the file upload functionality, I found some issues with the multipart requests. This fixes it .

christyjacob4 commented on templates/kotlin/src/main/java/io/appwrite/Client.kt.twig:
```diff
@@ -0,0 +1,232 @@
+                        val list = it.value as List<*>
+                        for (index in list.indices) {
+                            builder.addFormDataPart(
+                                "${it.key}[${index}]",
```
```suggestion
                                "${it.key}[]",
```
@abnegate  Not sure if the index is required here considering a cURL request does not require one. 

```curl
curl --location --request POST 'https://localhost/v1/storage/files' \
--header 'Content-Type: multipart/form-data' \
--header 'X-Appwrite-Key:  key' \
--header 'X-Appwrite-Project: project' \
--form 'file=@"/Users/test/Desktop/137-1377599_lady-bug-svg-cut-file-ladybug.png"' \
--form 'read[]="*"' \
--form 'write[]="*"' \
--form 'read[]="user:24er2f4we"'
```

christyjacob4 commented:
@abnegate  I am trying to test the OAuth stuff too. Will let you know if I have some progress. 
As @lohanidamodar suggested, you can disable clamav while testing. 

Also, could you update the [spec generation file](https://github.com/appwrite/sdk-generator/blob/e0a2a03ce49ccbb7c9cf59ddc5791eb958dd2d9e/src/SDK/Language/Kotlin.php) to include [these new files](https://github.com/appwrite/sdk-generator/pull/150/commits/2b5e11658dd30c5e65fb0ef4528b3ad4514d7d53) in the generated sdk ?

abnegate commented:
@lohanidamodar Thanks I'll give that a go soon

@christyjacob4 Woops missed that in the last commit, have added the oauth required files to the language file

christyjacob4 commented:
@abnegate  No problem :)

christyjacob4 commented:
@abnegate I seem to be having trouble with exception handling using viewModelScope. This is the code I am using

```kotlin

```

christyjacob4 commented on templates/kotlin/src/main/java/io/appwrite/Client.kt.twig:
```diff
@@ -0,0 +1,232 @@
+                response.use { res ->
+                    if (res.code >= 400) {
+                        val err = response.body?.string()?.fromJson(Error::class.java)
+                        throw AppwriteException("${err?.code}: ${err?.message}")
```
Okay, I got the issue. Instead of throwing the exception here, you need to cancel the job with the `cause` 
like so 
```suggestion
                        it.cancel(AppwriteException("${err?.code}: ${err?.message}"))
```

christyjacob4 commented:
@abnegate  Also noticed that the `httpBuilder`  is not being used anywhere 

```kotlin
/**
     * Get File for Download
     *
     * Get a file content by its unique ID. The endpoint response return with a
     * 'Content-Disposition: attachment' header that tells the browser to start
     * downloading the file to user downloads directory.
     *
     * @param fileId
     * @return The request response with a JSON body 
     */
    @JvmOverloads
    @Throws(AppwriteException::class)
    suspend fun getFileDownload(
		fileId: String
	): Response {
        val path = "/storage/files/{fileId}/download".replace("{fileId}", fileId)
        val params = mapOf<String, Any?>(
            "project" to client.config["project"],
            "key" to client.config["key"]
        )


        val httpBuilder = HttpUrl.Builder().build().newBuilder(client.endPoint + path)
            ?: throw AppwriteException("Badly formed url!")

        params.forEach {
            httpBuilder.addQueryParameter(it.key, it.value.toString())
        }

        return client.call("GET", path, params = params)
    }
```

abnegate commented:
@christyjacob4 Yep just saw that too, sorted now. File upload and download is working as expected for me now. Looking back into OAuth

christyjacob4 commented:
Awesome @abnegate could you share your file download code ? I was just working on that

abnegate commented:
@christyjacob4 

For an image:

```kotlin
        val downloadResponse = storageService.getFileDownload("file id")
        val image = BitmapFactory.decodeStream(downloadResponse.body!!.byteStream())
        binding.imageviewImage.setImageBitmap(image)
```

Generated from the latest commit:

```kotlin
    @JvmOverloads
    @Throws(AppwriteException::class)
    suspend fun getFileDownload(
		fileId: String
	): Response {
        val path = "/storage/files/{fileId}/download".replace("{fileId}", fileId)
        val params = mapOf<String, Any?>(
            "project" to client.config["project"],
            "key" to client.config["key"]
        )

        val stringParams = params.mapValues {
            if (it.value is Int || it.value is Double) {
                it.value.toString()
            }
            it.value
        }

        return client.call("GET", path, params = stringParams)
    }
```

christyjacob4 commented:
Thanks @abnegate  👍  

Is this necessary ? 

```kotlin
 val stringParams = params.mapValues {
            if (it.value is Int || it.value is Double) {
                it.value.toString()
            }
            it.value
        }
```

abnegate commented:
@christyjacob4 I'm not sure, that's just how it is in the flutter sdk so I had assumed so: 

https://github.com/appwrite/sdk-for-flutter/blob/aee8859e7f45c3b8b8569d2ae0eff5205a77a2c1/lib/services/storage.dart#L121

christyjacob4 commented:
@abnegate I think you should drop those. I believe Okhttp does this internally

abnegate commented:
@christyjacob4 Sweet removed the mapping, OAuth + cookie saving/sharing sorted now too, you'll need to change the activity definition in your manifest like so:

```xml
        <activity android:name="io.appwrite.views.CallbackActivity" >
            <intent-filter android:label="android_web_auth">
                <action android:name="android.intent.action.VIEW" />
                <category android:name="android.intent.category.DEFAULT" />
                <category android:name="android.intent.category.BROWSABLE" />
                <data android:scheme="appwrite-callback-[PROJECT_ID]"  />
            </intent-filter>
        </activity>
```

christyjacob4 commented:
@abnegate this is top notch 👌
I will test it in a while and let you know

christyjacob4 commented:
@abnegate  Great work! Just tested the OAuth Flow. Works like a charm 👏  Great Work
Sorry for all these critical reviews 😅

Just one more thing I guess, Can you also add some docs to the extra classes that we have created for the OAuth Callback and stuff ? 

Then let's add the oAuth instructions to the readme, just like the flutter SDK and we can merge this

TorstenDittmann commented on tests/languages/kotlin/ServiceTest.kt:
```diff
@@ -0,0 +1,69 @@
+        // General Tests
+        response = general.redirect().execute()
+        printResponse(response)
+
```
I think we are missing the Exception tests here 🙂

TorstenDittmann reviewed (approved):
Awesome work @abnegate 👏🏻 

Really appreciated!

abnegate commented:
@christyjacob4 Added the docs, wasn't so sure about how to update the readme, looks like should be done with `sdk.setGettingStarted` but there was no other references to that

christyjacob4 commented:
@abnegate  Yes, you're right. The example.php which we use only for development does not have the getting started section. 

Take a look at the production script that we use to generate the SDKs where we set all the params [here](https://github.com/appwrite/appwrite/blob/master/app/tasks/sdks.php#L192)

christyjacob4 commented on example.php:
````diff
@@ -398,6 +399,110 @@ function getSSLPage($url) {
+```
+
+### Learn more
+You can use followng resources to learn more and get help
````
```suggestion
You can use following resources to learn more and get help
```

christyjacob4 commented:
@abnegate Just some more minor changes. Please address them 😄 

1. Rename `AppwriteClient.kt` to `Client.kt`
2. Let's stick to lower case header names [here](https://github.com/christyjacob4/appwrite-android-demo/blob/main/kotlin-android/src/main/java/io/appwrite/AppwriteClient.kt#L63)  
3. The service names need not have a `Service` suffix [here](https://github.com/christyjacob4/appwrite-android-demo/tree/main/kotlin-android/src/main/java/io/appwrite/services). Let's keep them as `Account.kt` `Avatars.kt` etc. 
4. Can we keep the indentation for headers consistent with that of params?  [here](https://github.com/christyjacob4/appwrite-android-demo/blob/main/kotlin-android/src/main/java/io/appwrite/services/AccountService.kt#L116-L121)
5. Let's leave a line break before the return statement [here](https://github.com/christyjacob4/appwrite-android-demo/blob/main/kotlin-android/src/main/java/io/appwrite/services/AccountService.kt#L196)
6. Let's remove the [Error class](https://github.com/christyjacob4/appwrite-android-demo/tree/main/kotlin-android/src/main/java/io/appwrite/models ) as it looks like this is not being used anywhere .

That's the end of it 😅  
Again, amazing work @abnegate  !!!

christyjacob4 commented:
@abnegate Let us know if you need any help with this 😄

abnegate commented:
@christyjacob4 Should be good now let me know if I've missed anything

christyjacob4 commented:
@abnegate My sincere apologies, the Error class was indeed being user [here](https://github.com/christyjacob4/appwrite-android-demo/blob/main/kotlin-android/src/main/java/io/appwrite/AppwriteClient.kt#L11) and [here](https://github.com/christyjacob4/appwrite-android-demo/blob/main/kotlin-android/src/main/java/io/appwrite/AppwriteClient.kt#L240)

Can we please add that back

christyjacob4 commented:
@abnegate 🥳  
Thank you so much for sticking around with the feedback loop! 
We really appreciate your help with this feature. ❤️

christyjacob4 reviewed (approved):
Amazing Job @abnegate  💯

kutear commented:
it's ready to release?  @christyjacob4

christyjacob4 commented:
@kutear yes this is ready for a beta release. You can check out its usage in this demo app https://github.com/christyjacob4/appwrite-android-demo

eldadfux commented:
@abnegate amazing contribution!! ❤️ 🥳

christyjacob4 commented:
Hey @abnegate  

We need some help with the deployment of the sdk. 
We considered 4 options for the deployment channel 
1. bintray and jcenter ❌ 
2. Github Packages ❌ 
3. Jitpack ❌ 
4. Maven Central ✅  

As you might have seen, jcenter is being sunset and it also had a lot of [security issues](https://blog.autsoft.hu/a-confusing-dependency/)
Github packages requires authentication to both read and write packages
Jitpack is not very scalable in the long run.

This leaves us with MavenCentral as the only viable option. We are in the process of creating a mavencentral account . Meanwhile I tried to use the deploy the library to a local maven repo and it didn't seem to work. Could you give it another look at your end? 

Secondly, we are also looking for ways to deploy the same library to JVM as well for our server side SDK. I'm unfamiliar with this and it would be great if you could help 😊
